[pytorch][doc] Threading and CPU Inference note

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23417 [pytorch][doc] Threading and CPU Inference note**

A note on thread settings for inference.

cd docs; make html

Differential Revision: [D16523781](https://our.internmc.facebook.com/intern/diff/D16523781)

Commit:
[pytorch][doc] Threading and CPU Inference note

Summary:
A note on thread settings for inference.

Test Plan:
cd docs; make html

Commit:
Update on "[pytorch][doc] Threading and CPU Inference note"


Summary:
A note on thread settings for inference.

Test Plan:
cd docs; make html

ilia-cher commented:
[html.zip](https://github.com/pytorch/pytorch/files/3434201/html.zip)

soumith commented on docs/source/notes/inference_threading.rst:
```diff
@@ -0,0 +1,86 @@
+.. _inference-threading:
```
the heading should be cpu inference as well, if it shows up in the side bar

Commit:
Update on "[pytorch][doc] Threading and CPU Inference note"


Summary:
A note on thread settings for inference.

Test Plan:
cd docs; make html

ilia-cher commented:
Would appreciate comments on wording too
[html.zip](https://github.com/pytorch/pytorch/files/3437324/html.zip)

soumith commented on docs/source/notes/threading_cpu_inference.rst:
```diff
@@ -0,0 +1,88 @@
+
+PyTorch supports OpenMP and TBB implementations of intra-op parallelism specified
+with the build settings:
+
```
Add a pragraph stating why you would want to use multiple libraries and what trade-offs they might have -- or why the user should be interested in these build options / build settings.

ilia-cher commented on docs/source/notes/threading_cpu_inference.rst:
```diff
@@ -0,0 +1,88 @@
+One or more inference threads execute model's forward pass on the given inputs.
+Each inference thread invokes a JIT interpreter that executes the ops
+of a model inline, one by one. A model can utilize a ``fork`` TorchScript
+primitive to launch some of the ops in parallel, as an asynchronous task, and
```
sounds good, the only minor change - I wanted to avoid explicit 'you' here

ilia-cher commented on docs/source/notes/threading_cpu_inference.rst:
```diff
@@ -0,0 +1,88 @@
+primitive to launch some of the ops in parallel, as an asynchronous task, and
+synchronize with them later, using a future returned from ``fork``.
+
+PyTorch uses a single thread pool for the inter-op parallelism, this thread pool
```
minor correction: the tasks don't have to be forked at the same time, the thread pool is going to be the same in any case

Commit:
Update on "[pytorch][doc] Threading and CPU Inference note"


Summary:
A note on thread settings for inference.

Test Plan:
cd docs; make html

Differential Revision: [D16523781](https://our.internmc.facebook.com/intern/diff/D16523781)

ilia-cher commented:
Partial rewrite based on @soumith comments
[html_doc.zip](https://github.com/pytorch/pytorch/files/3437903/html_doc.zip)

apaszke commented on docs/source/notes/cpu_threading_torchscript_inference.rst:
```diff
@@ -0,0 +1,126 @@
+The following figure shows different levels of parallelism one would find in a
+typical application:
+
+.. image:: cpu_threading_torchscript_inference.png
```
Can we make this a vector graphic? 78kB tucked to the repo history forever is not a idea 😕

jspisak commented:
@brianjo - does this PR need to be in a different repo?

brianjo commented:
> @brianjo - does this PR need to be in a different repo?

Should be good once the branch gets merged.

Commit:
Update on "[pytorch][doc] Threading and CPU Inference note"


Summary:
A note on thread settings for inference.

Test Plan:
cd docs; make html

Differential Revision: [D16523781](https://our.internmc.facebook.com/intern/diff/D16523781)

ilia-cher commented:
Using SVG, minor edits.

ilia-cher commented:
[html_doc.zip](https://github.com/pytorch/pytorch/files/3444379/html_doc.zip)

Commit:
Update on "[pytorch][doc] Threading and CPU Inference note"


Summary:
A note on thread settings for inference.

Test Plan:
cd docs; make html

Differential Revision: [D16523781](https://our.internmc.facebook.com/intern/diff/D16523781)
